feat: HUMSIG-37: Enhancing ML backend connection experience

niklub opened the pull request:
### PR fulfills these requirements
- [ ] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [ ] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [ ] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [ ] Self-reviewed and ran all changes on a local instance (for bug fixes/features)



#### Change has impacts in these area(s)
_(check all that apply)_
- [ ] Product design
- [ ] Backend (Database)
- [ ] Backend (API)
- [ ] Frontend



### Describe the reason for change
_(link to issue, supportive screenshots etc.)_



#### What does this fix?
_(if this is a bug fix)_



#### What is the new behavior?
_(if this is a breaking or feature change)_



#### What is the current behavior?
_(if this is a breaking or feature change)_



#### What libraries were added/updated?
_(list all with version changes)_



#### Does this change affect performance?
_(if so describe the impacts positive or negative)_



#### Does this change affect security?
_(if so describe the impacts positive or negative)_



#### What alternative approaches were there?
_(briefly list any if applicable)_



#### What feature flags were used to cover this change?
_(briefly list any if applicable)_



### Does this PR introduce a breaking change?
_(check only one)_
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [ ] No
- [ ] Not sure (briefly explain the situation below)



### What level of testing was included in the change?
_(check all that apply)_
- [ ] e2e
- [ ] integration
- [ ] unit



### Which logical domain(s) does this change affect?
_(for bug fixes/features, be as precise as possible. ex. Authentication, Annotation History, Review Stream etc.)_


- https://github.com/HumanSignal/dm2/pull/310

Commit:
Updated implementation of the ML Backend experience. Docs & Test to follow

Commit:
Updating the experience a bit more based on the feedback. Plus
updating tests

Commit:
minor updates on the experience

Commit:
removing debug info, uncommenting bits

deppp committed:
Merge branch 'develop' into dev/ml-backend-exp

Commit:
fixing small issue to make it backward compatible with the previous project setting

Commit:
Merge branch 'dev/ml-backend-exp' of https://github.com/heartexlabs/label-studio into dev/ml-backend-exp

Commit:
Fix errors, code cleanup, fix ruff

Commit:
Downgrade testing-library, reformat with linters

Commit:
Remove excessive calls in task api

Commit:
Fix frontend script

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8192397383

jombooth committed:
fmt

bmartel commented on web/apps/labelstudio/src/components/Form/Elements/Select/Select.js:
```diff
@@ -4,11 +4,30 @@ import { FormField } from '../../FormField';
+    return groupedOptions;
+  }, {});
+
+  const renderOption = (option) => {
+    const value = option.value ?? option;
+    const label = option.label ?? value;
+    const disabled = option.disabled ?? false;
+    const hidden = option.hidden ?? false;
+
+    return (
+      <option key={value} value={value} disabled={disabled} hidden={hidden}>
+        {label}
+      </option>
+    );
+  };
```
I would move this up to its own separate component definition outside of `Select` to allow less logic to be defined within another.

```suggestion
  const SelectOption = ({ value, label, disabled = false, hidden = false, ...props }) => {
    return (
      <option value={value} disabled={disabled} hidden={hidden} {...props}>
        {label ?? value}
      </option>
    );
  };
```

bmartel commented on web/apps/labelstudio/src/components/Form/Elements/Select/Select.js:
```diff
@@ -41,19 +60,14 @@ const Select = ({ label, className, options, validate, required, skip, labelProp
+	      {
+		Object.keys(grouped).map((group) => {		
+		  return (group === 'NoGroup') ?
+		    grouped[group].map(renderOption) :
```
```suggestion
		    grouped[group].map((option) => <SelectOption {...(option.value ? { ...option, key: option.value } : { value: option, key: option })} />) :
```

bmartel commented on web/apps/labelstudio/src/components/Form/Elements/Select/Select.js:
```diff
@@ -41,19 +60,14 @@ const Select = ({ label, className, options, validate, required, skip, labelProp
+		Object.keys(grouped).map((group) => {		
+		  return (group === 'NoGroup') ?
+		    grouped[group].map(renderOption) :
+		    <optgroup label={group}> {grouped[group].map(renderOption)} </optgroup>
```
```suggestion
		    <optgroup label={group}> {grouped[group].map((option) => <SelectOption {...(option.value ? { ...option, key: option.value } : { value: option, key: option })} />)} </optgroup>
```

bmartel commented on web/apps/labelstudio/src/config/ApiConfig.js:
```diff
@@ -53,8 +53,9 @@ export const API_CONFIG = {
     deleteMLBackend: "DELETE:/ml/:pk",
     trainMLBackend: "POST:/ml/:pk/train",
-    predictWithML: "POST:/ml/:pk/predict",
+    predictWithML: "POST:/ml/:pk/predict",    
```
```suggestion
    predictWithML: "POST:/ml/:pk/predict",
```

The linters should be taking care of this very soon!

bmartel commented on web/apps/labelstudio/src/components/Form/Elements/ToggleRight/ToggleRight.styl:
```diff
@@ -0,0 +1,56 @@
+.toggle-right
```
@nicholasrq You mentioned SCSS is the way forward, how strongly are we taking that at this time with regards to new code? With Stylus having been long deprecated we are going to have to move all existing Stylus over to SCSS, and wanted to know if that is something we are trying to encourage in new PRs?

bmartel commented on web/apps/labelstudio/src/assets/icons/index.js:
```diff
@@ -26,5 +26,8 @@ export { ReactComponent as LsThumbsUp } from './thumbs-up.svg';
 export { ReactComponent as IconUpload } from './upload.svg';
 export { ReactComponent as LsPencil } from './pencil.svg';
 export { ReactComponent as IconInfoOutline } from './info-outline.svg';
+export { ReactComponent as IconSparkGrey } from './spark_grey.svg'
```
Can we rename this to kebab-case? The other surrounding definitions are using that. (locality nit, not a preference of mine 🤣)
```suggestion
export { ReactComponent as IconSparkGrey } from './spark-grey.svg'
```

bmartel commented on web/apps/labelstudio/src/assets/icons/spark_grey.svg:
We should keep the locality convention of kebab for these files:

`web/apps/labelstudio/src/assets/icons/spark-grey.svg`

bmartel commented on web/apps/labelstudio/src/components/Form/Form.js:
```diff
@@ -169,16 +169,14 @@ export default class Form extends React.Component {
-
-          if (['checkbox', 'radio'].includes(field.type)) {
+          const inputValue = field.value;	    
+            if (['checkbox', 'radio'].includes(field.type)) {
```
Is this a formatting blip?

bmartel commented on web/libs/datamanager/src/components/DataManager/Toolbar/ActionsButton.js:
```diff
@@ -45,14 +45,14 @@ export const ActionsButton = injector(observer(({ store, size, hasSelected, ...r
   const invokeAction = (action, destructive) => {
     if (action.dialog) {
-      const { type: dialogType, text, form } = action.dialog;
+      const { type: dialogType, text, form, title, modal_title } = action.dialog;
```
In the below definition of the ActionDialog model it is missing `title`. Do we need both `title` and `modal_title`?

If not, we can collapse this down to:

```suggestion
      const { type: dialogType, text, form, title } = action.dialog;
```

If we do require both I'd suggest camelCase'ing the modal_title to be locally consistent:

```suggestion
      const { type: dialogType, text, form, title, modalTitle } = action.dialog;
```

bmartel commented on web/libs/datamanager/src/components/DataManager/Toolbar/ActionsButton.js:
```diff
@@ -45,14 +45,14 @@ export const ActionsButton = injector(observer(({ store, size, hasSelected, ...r
+      
       dialog({
-        title: destructive ? "Destructive action." : "Confirm action.",
+        title: modal_title ? modal_title : destructive ? "Destructive action" : "Confirm action",
```
If distinguishing between `title` and `modal_title`:
```suggestion
        title: modalTitle ? modalTitle : destructive ? "Destructive action" : "Confirm action",
```

bmartel commented on web/libs/datamanager/src/stores/Action.js:
```diff
@@ -44,6 +44,7 @@ const ActionDialog = types.model("ActionDialog", {
   text: types.string,
   type: types.enumeration(["confirm", "prompt"]),
   form: types.maybeNull(types.array(ActionFormRow)),
+  modal_title: types.maybeNull(types.string),
```
I think this is ideal here, as we already have distinguished it is an ActionDialog.
```suggestion
  title: types.maybeNull(types.string),
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/PredictionsSettings/PredictionsList.js:
```diff
@@ -0,0 +1,89 @@
+  );
+};
+
+const VersionCard = ({ version, selected, onSelect, edittable, onDelete }) => {
```
```suggestion
const VersionCard = ({ version, selected, onSelect, editable, onDelete }) => {
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/PredictionsSettings/PredictionsList.js:
```diff
@@ -0,0 +1,89 @@
+const VersionCard = ({ version, selected, onSelect, edittable, onDelete }) => {
+    const rootClass = cn("predictionCard");
+    const disabledStyle = {
+        opacity: edittable ? "1" : 0.3
```
```suggestion
        opacity: editable ? "1" : 0.3
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/PredictionsSettings/PredictionsSettings.js:
```diff
@@ -0,0 +1,94 @@
+  const [activeVersion, setActiveVersion] = useState();
+  // const [editEnable, setEditEnable] = useState();
+
+  const edittable = ! project.evaluate_predictions_automatically;
```
```suggestion
  const editable = ! project.evaluate_predictions_automatically;
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/PredictionsSettings/PredictionsSettings.js:
```diff
@@ -0,0 +1,94 @@
+  const [loading, setLoading] = useState(false);
+  const [loaded, setLoaded] = useState(false);
+  const [activeVersion, setActiveVersion] = useState();
+  // const [editEnable, setEditEnable] = useState();
```
Is this meant to be used?
```suggestion
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/PredictionsSettings/PredictionsSettings.styl:
```diff
@@ -0,0 +1,95 @@
+
+.pred-settings
```
```suggestion
.prediction-settings
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/PredictionsSettings/PredictionsSettings.js:
```diff
@@ -0,0 +1,94 @@
+  }, [project]);
+
+    return (
+      <Block name="pred-settings">
```
```suggestion
      <Block name="prediction-settings">
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/PredictionsSettings/PredictionsList.js:
```diff
@@ -0,0 +1,89 @@
+
+  
+    return (
+      <Block name="predictionCard">
```
Just to keep consistent with the remainder of what I see in this PR. This is another job for a linter, as I have seen many takes on which way to go, which as long as we are consistent its correct 🙂. (I'm going to be pushing that we put in place consistent linting tools across the MonoRepos be put in place very soon so PRs will just not have to worry about such things.)

```suggestion
      <Block name="prediction-card">
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/PredictionsSettings/PredictionsList.styl:
```diff
@@ -0,0 +1,42 @@
+.predictionCard
```
```suggestion
.prediction-card
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/PredictionsSettings/PredictionsList.js:
```diff
@@ -0,0 +1,89 @@
+};
+
+const VersionCard = ({ version, selected, onSelect, edittable, onDelete }) => {
+    const rootClass = cn("predictionCard");
```
```suggestion
    const rootClass = cn("prediction-card");
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/TestRequest.js:
```diff
@@ -0,0 +1,55 @@
+      </Caption>
+      <Elem name={'blocks'}>
+        <Elem name={'left'}>
+          {/* <h4>Task</h4> */}
+          {/* <Elem name={'code'} style={{ maxHeight: "100px" }}><pre>{JSON.stringify(testResponse.task, null, 2)}</pre></Elem> */}
```
Is this part still needed?
```suggestion
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/StartModelTraining.js:
```diff
@@ -0,0 +1,45 @@
+  return (
+    <Block name="test-request">
+      <Description style={{ marginTop: 0, maxWidth: 680 }}>
+        You're about to manually trigger your model's training process. This action will start the learning phase based on how train method is implemented in the ML Backend. Proceed to begin this process.<br /><br />
```
```suggestion
        You're about to manually trigger your model's training process. This action will start the learning phase based on how the train method is implemented in the ML Backend. Proceed to begin this process.<br /><br />
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/MachineLearningSettings.js:
```diff
@@ -4,192 +4,151 @@ import { Description } from '../../../components/Description/Description';
+            body: <StartModelTraining backend={backend} />
+        };
+        
+        const modalRef = modal(modalProps);
```
This doesn't appear to be referenced in scope.

```suggestion
```

bmartel replied:
This doesn't appear to be referenced within scope, so we can drop it down to:
```suggestion
        modal(modalProps);
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/MachineLearningSettings.js:
```diff
@@ -4,192 +4,151 @@ import { Description } from '../../../components/Description/Description';
+            body: <TestRequest backend={backend} />
+        };
+        
+        const modalRef = modal(modalProps);
```
```suggestion
        modal(modalProps);
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/MachineLearningList.js:
```diff
@@ -56,43 +52,42 @@ const BackendCard = ({ backend, onStartTrain, onEdit, onDelete }) => {
-      </Button>
-    </Card>
-  );
+  const rootClass = cn("backendCard");
```
```suggestion
  const rootClass = cn("backend-card");
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/MachineLearningList.js:
```diff
@@ -56,43 +52,42 @@ const BackendCard = ({ backend, onStartTrain, onEdit, onDelete }) => {
+  const rootClass = cn("backendCard");
+    
+    return (
+        <Block name="backendCard">
```
```suggestion
        <Block name="backend-card">
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/MachineLearningList.styl:
```diff
@@ -0,0 +1,30 @@
+
+.backendCard
```
```suggestion
.backend-card
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/MachineLearningList.styl:
```diff
@@ -0,0 +1,30 @@
+  background #FAFAFA
+  border 1px solid rgba(137, 128, 152, 0.16)
+
+  &__titleBlk
```
```suggestion
  &__title-container
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/MachineLearningList.js:
```diff
@@ -56,43 +52,42 @@ const BackendCard = ({ backend, onStartTrain, onEdit, onDelete }) => {
+    
+    return (
+        <Block name="backendCard">
+          <div className={rootClass.elem("titleBlk")}>
```
```suggestion
          <div className={rootClass.elem("title-container")}>
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/Forms.js:
```diff
@@ -0,0 +1,97 @@
+          onSubmit={async (response) => {
+            if (!response.error_message) {
+              onSubmit(response);
+              /* await fetchBackends(); */
+              /* modalRef.close(); */
```
Are these needed still?
```suggestion
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/Forms.js:
```diff
@@ -0,0 +1,97 @@
+        <Input name="url" label="Backend URL" required />
+      </Form.Row>
+
+      {/* <h4>Authentication Method</h4> */}
```
Is this still needed?
```suggestion
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/Forms.js:
```diff
@@ -0,0 +1,97 @@
+      </Form.Row>
+      
+      <Form.Row columnCount={1}>
+        <div>
+          <Toggle
+            name="is_interactive"
+            label="Interactive preannotations"
+            description="If enabled some labeling tools will send requests to the ML Backend interactively during the annotation process."
+          />
+        </div>
```
Do we need the wrapping div here? Not sure of the Form.Row behaviour, but this might be to protect against flex grow caused by the parent. If this is necessary please disregard the suggestion.

```suggestion
        <Toggle
          name="is_interactive"
          label="Interactive preannotations"
          description="If enabled some labeling tools will send requests to the ML Backend interactively during the annotation process."
        />
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/AnnotationSettings/ModelVersionSelector.js:
```diff
@@ -0,0 +1,97 @@
+      },
+    });
+
+    if (modelVersions) {
+      if (modelVersions.live.length > 0) {
+        
+        const liveModels = modelVersions.live.map(item => {
+          const label = item.title + " (" + item.readable_state + ")";
+          return {
+            group: "Models",
+            value: item.title,
+            label: label
+          };
+        });
+        
+        setModels(liveModels);
+      }
+
+      if (modelVersions.static.length > 0) {
+        const staticModels = modelVersions.static.map(item => {
+          const label = item.model_version + " (" + item.count + " predictions)";
+          return {
+            group: "Predictions",
+            value: item.model_version,
+            label: label
+          };
+        });
+        
+        setVersions(staticModels);
+      }
+
+      if (!(modelVersions.static?.length) && !(modelVersions.live?.length)) {
+        setPlaceholder("No model or predictions available");
+      }      
+    }
```
Can use null safe checks to remove the additional wrapping if statement:

```suggestion
    if (modelVersions?.live?.length > 0) {
      const liveModels = modelVersions.live.map(item => {
        const label = item.title + " (" + item.readable_state + ")";
        return {
          group: "Models",
          value: item.title,
          label: label
        };
      });
      
      setModels(liveModels);
    }

    if (modelVersions?.static?.length > 0) {
      const staticModels = modelVersions.static.map(item => {
        const label = item.model_version + " (" + item.count + " predictions)";
        return {
          group: "Predictions",
          value: item.model_version,
          label: label
        };
      });
      
      setVersions(staticModels);
    }
  
    if (!(modelVersions?.static?.length) && !(modelVersions.live?.length)) {
      setPlaceholder("No model or predictions available");
    }      
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/AnnotationSettings.js:
```diff
@@ -0,0 +1,88 @@
+          </div>
+          <div style={{color: "rgba(0,0,0,0.4)" }}>
+            <p>Write instructions to help users complete labeling tasks.</p>
+            <p>The instruction field supports HTML markup and it allows to integrate with images, iframes (pdf).</p>
```
```suggestion
            <p>The instruction field supports HTML markup and it allows use of images, iframes (pdf).</p>
```

bmartel commented on web/libs/datamanager/src/components/DataManager/Toolbar/ActionsButton.js:
As the MonoRepo PRs have yet to all close in DM/LSF/LSP, we are going to be allowing those to merge/close up until the start of the next sprint, which will then be operating here strictly as the only source of truth. 

To avoid this getting overridden by those other PRs, the DM changes made here will need to be made to the old repository by opening a PR on the branch of the same name: `fb-ml-backend-exp` in https://github.com/HumanSignal/dm2

If this PR will merge later than a week from now, then it will not require any further action.

bmartel replied:
Setup the branch, should be now linked as well.

bmartel commented on web/package.json:
```diff
@@ -127,7 +127,7 @@
     "@swc/cli": "~0.1.62",
     "@swc/core": "1.3.96",
-    "@testing-library/react": "14.0.0",
+    "@testing-library/react": "12",
```
Is this version change intended to go down?

niklub replied:
The current unit tests are not compatible with this version

bmartel reviewed (changes requested):
Some minor changes suggested, a few questions, but logically sound.

Commit:
fix sdk version & change api /predict/test

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8201028879

Commit:
try update lock

bmartel committed:
Merge branch 'develop' into 'fb-ml-backend-exp'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8207748036

bmartel committed:
[submodules] Copy src HumanSignal/dm2

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8207748036

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8207796735

Commit:
Address review comments with stylistic changes, remove unusable code, change some writings

Commit:
Additional changes

Commit:
Add svg icon file

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8217959028

Commit:
Fix ml/predict/test api

bmartel commented on web/apps/labelstudio/src/pages/Settings/AnnotationSettings/ModelVersionSelector.js:
```diff
@@ -0,0 +1,94 @@
+import {ProjectContext} from '../../../providers/ProjectProvider';
+
+
+export const ModelVersionSelector = ({
+                                       name = "model_version",
+                                       valueName = "model_version",
+                                       apiName = "projectModelVersions",
+                                       ...props
+                                     }) => {
```
```suggestion
export const ModelVersionSelector = ({
  name = "model_version",
  valueName = "model_version",
  apiName = "projectModelVersions",
  ...props
}) => {
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/AnnotationSettings.js:
```diff
@@ -0,0 +1,81 @@
+                      name="show_collab_predictions"
+                      onChange={(e) => {
+                        setCollab(e.target.checked)
+                      }}/>
```
```suggestion
              <Toggle 
                label="Use predictions to pre-label data"
                description="Enable and select which set of predictions to use for prelabeling."
                name="show_collab_predictions"
                onChange={(e) => {
                  setCollab(e.target.checked)
                }}
              />
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/Forms.js:
```diff
@@ -0,0 +1,92 @@
+  const [mlError, setMLError] = useState();
+  
+  return (
+    <Form action={action}
+          formData={{ ...(backend ?? {}) }}
+          params={{ pk: backend?.id }}
+          onSubmit={async (response) => {
+            if (!response.error_message) {
+              onSubmit(response);
+            }
+          }}>
```
```suggestion
    <Form action={action}
      formData={{ ...(backend ?? {}) }}
      params={{ pk: backend?.id }}
      onSubmit={async (response) => {
        if (!response.error_message) {
          onSubmit(response);
        }
      }}>
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/Forms.js:
```diff
@@ -0,0 +1,92 @@
+      </Form.Row>
+      
+      <Form.Row columnCount={1}>
+          <Toggle
+            name="is_interactive"
+            label="Interactive preannotations"
+            description="If enabled some labeling tools will send requests to the ML Backend interactively during the annotation process."
+          />
```
```suggestion
        <Toggle
          name="is_interactive"
          label="Interactive preannotations"
          description="If enabled some labeling tools will send requests to the ML Backend interactively during the annotation process."
        />
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/MachineLearningList.js:
```diff
@@ -56,43 +52,42 @@ const BackendCard = ({ backend, onStartTrain, onEdit, onDelete }) => {
+            </div>
+            <div className={rootClass.elem("menu")}>
+              <Dropdown.Trigger align="right" content={(
+                <Menu size="medium" contextual>
+                    <Menu.Item onClick={() => onTestRequest(backend)}>Send Test Request</Menu.Item>
+                    <Menu.Item onClick={() => onStartTrain(backend)}>Start Training</Menu.Item>
+                    
+                    <Menu.Item onClick={() => onEdit(backend)}>Edit</Menu.Item>
+                  <Menu.Item onClick={() => confirmDelete(backend)} isDangerous>Delete</Menu.Item>
+                  </Menu>
```
```suggestion
                <Menu size="medium" contextual>
                  <Menu.Item onClick={() => onTestRequest(backend)}>Send Test Request</Menu.Item>
                  <Menu.Item onClick={() => onStartTrain(backend)}>Start Training</Menu.Item>
                  <Menu.Item onClick={() => onEdit(backend)}>Edit</Menu.Item>
                  <Menu.Item onClick={() => confirmDelete(backend)} isDangerous>Delete</Menu.Item>
                </Menu>
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/MachineLearningSettings.js:
```diff
@@ -1,195 +1,152 @@
+      title: `${backend ? 'Edit' : 'Connect'} Model`,
+      style: {width: 760},
+      closeOnClickOutside: false,
+      body: <CustomBackendForm action={action}
+                               backend={backend}
+                               project={project}
+                               onSubmit={() => {
+                                 fetchBackends();
+                                 modalRef.close();
+                               }}/>
```
```suggestion
      body: (
        <CustomBackendForm
          action={action}
          backend={backend}
          project={project}
          onSubmit={() => {
            fetchBackends();
            modalRef.close();
          }} 
        />
      )
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/MachineLearningSettings.js:
```diff
@@ -1,195 +1,152 @@
+    <Block name="ml-settings">
+      <Elem name={'wrapper'}>
+        {loading && <Spinner size={32}/>}
+        {loaded && backends.length == 0 &&
+          <EmptyState icon={<IconEmptyPredictions/>}
+                      title="Let’s connect your first model"
+                      description="Connect a machine learning model to generate predictions. These predictions can be compared side by side, used for efficient pre‒labeling and, to aid in active learning, directing users to the most impactful labeling tasks."
+                      action={<Button primary onClick={() => showMLFormModal()}>Connect Model</Button>}
+                      footer={<div>Need help?<br/><a>Learn more about connecting models in our docs</a></div>}/>}
```
```suggestion
        {loaded && backends.length == 0 && (
          <EmptyState icon={<IconEmptyPredictions/>}
            title="Let’s connect your first model"
            description="Connect a machine learning model to generate predictions. These predictions can be compared side by side, used for efficient pre‒labeling and, to aid in active learning, directing users to the most impactful labeling tasks."
            action={<Button primary onClick={() => showMLFormModal()}>Connect Model</Button>}
            footer={<div>Need help?<br/><a>Learn more about connecting models in our docs</a></div>}
          />
        )}
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/MachineLearningSettings.js:
```diff
@@ -1,195 +1,152 @@
+                      description="Connect a machine learning model to generate predictions. These predictions can be compared side by side, used for efficient pre‒labeling and, to aid in active learning, directing users to the most impactful labeling tasks."
+                      action={<Button primary onClick={() => showMLFormModal()}>Connect Model</Button>}
+                      footer={<div>Need help?<br/><a>Learn more about connecting models in our docs</a></div>}/>}
+        <MachineLearningList onEdit={(backend) => showMLFormModal(backend)}
+                             onTestRequest={(backend) => showRequestModal(backend)}
+                             onStartTraining={(backend) => startTrainingModal(backend)}
+                             fetchBackends={fetchBackends}
+                             backends={backends}/>
```
```suggestion
        <MachineLearningList
          onEdit={(backend) => showMLFormModal(backend)}
          onTestRequest={(backend) => showRequestModal(backend)}
          onStartTraining={(backend) => startTrainingModal(backend)}
          fetchBackends={fetchBackends}
          backends={backends}
        />
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/MachineLearningSettings/StartModelTraining.js:
```diff
@@ -0,0 +1,47 @@
+
+        *Note: Currently, there is no built-in feedback loop within this interface for tracking the training progress.
+        You'll need to monitor the model's training steps directly through the model's own tools and environment.
+
+
```
```suggestion
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/PredictionsSettings/PredictionsList.js:
```diff
@@ -0,0 +1,89 @@
+
+  return (
+    <div style={{maxWidth: 680}}>
+      {versions.map(v => <VersionCard key={v.model_version}
+                                      version={v}
+                                      onDelete={onDelete}/>)}
```
```suggestion
      {versions.map(v => (
        <VersionCard
          key={v.model_version}
          version={v}
          onDelete={onDelete}
        />
      ))}
```

bmartel commented on web/apps/labelstudio/src/pages/Settings/index.js:
```diff
@@ -17,8 +18,9 @@ export const MenuLayout = ({ children, ...routeProps }) => {
         GeneralSettings,
         LabelingSettings,
-        InstructionsSettings,
+	      AnnotationSettings,
```
```suggestion
	AnnotationSettings,
```

bmartel committed:
Running a formatter/lint on the code

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8237179655

bmartel committed:
Running a formatter/lint on the code

bmartel committed:
Merge remote-tracking branch 'origin/fb-ml-backend-exp' into fb-ml-backend-exp

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8237277894

bmartel committed:
Running a formatter/lint on the code

bmartel committed:
Merge remote-tracking branch 'origin/fb-ml-backend-exp' into fb-ml-backend-exp

bmartel committed:
Running a formatter/lint on the code

bmartel reviewed (approved):
Ran a formatter/linter on the other frontend files, pushed that up as well. Looks good overall now.

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8237418085

Commit:
Handle security on password input

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8238405264

Commit:
Display default password when not specified

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8240219216

Commit:
Reduce number of calls due to project id cache

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8240697575

jombooth commented on label_studio/tasks/serializers.py:
```diff
@@ -711,13 +658,19 @@ def get_unique_lock_id(self, task):
-        else:
-            for ml_backend in project.ml_backends.all():
-                ml_backend.predict_tasks([task])
-            return super().get_predictions(task)
+        """ """
+        predictions = task.get_predictions_for_prelabeling()
+        return PredictionSerializer(predictions, many=True, read_only=True, default=[], context=self.context).data
+
+    # def get_predictions(self, task):
+    #     project = task.project
+
+    #     if not project.show_collab_predictions:
+    #         return []
+    # else:
+    #     # for ml_backend in project.ml_backends.all():
+    #     #     ml_backend.predict_tasks([task])
+    #     return super().get_predictions(task)
```
Need to make sure we didn't break anything with this change, and remove commented out code

bmartel commented on label_studio/data_manager/actions/basic.py:
```diff
@@ -138,6 +138,7 @@ def async_project_summary_recalculation(tasks_ids_list, project_id):
         'title': 'Retrieve Predictions',
         'order': 90,
         'dialog': {
+            'modal_title': 'Retrieve Predictions',
```
`modal_title` -> `title`

bmartel commented on label_studio/data_manager/actions/predictions_to_annotations.py:
```diff
@@ -95,8 +95,10 @@ def predictions_to_annotations_form(user, project):
         'dialog': {
-            'text': 'This action will create new annotations from predictions with the selected model version '
-            'for each selected task.',
+            'modal_title': 'Create Annotations From Predictions',
```
`modal_title` -> `title`
